### PR TITLE
Migrating to gradle: Add gradle shadow plugin

### PR DESCRIPTION
**Pull request by shashachu**
* Integrated https://github.com/johnrengelman/shadow
* `./gradlew clean build shadowJar` outputs `ktlint/build/ktlint-all.jar`
* `java -jar ktlint/build/ktlint-all.jar` works

### CODE REVIEW

**Comment by shashachu** [comment]
actually doesn't seem to be working properly

File: build.gradle [hunk 1]
```diff
@@ -44,6 +46,11 @@ configurations {
 
 dependencies {
   ktlint project(':ktlint')
+  subprojects.forEach { subproject ->
+    if (subproject.name != 'ktlint-ruleset-template') {
```
Comment (shashachu) [inline build.gradle]: This is pretty dopey, but excluding the project in the `shadowJar` config as described in this issue https://github.com/johnrengelman/shadow/issues/39 doesn't seem to work anymore.

**Comment by shashachu** [comment]
@jaredsburrows thoughts? This is cobbled together from some shadow plugin examples.

File: build.gradle (hunk 1 above)
Comment (rom4ek) [reply build.gradle]: But what's the reasoning behind including the template module into the project? I mean `settings.gradle`

Comment (shashachu) [reply build.gradle]: Presumably to ensure it keeps building, but maybe it just causes more issues than it solves.

Comment (rom4ek) [reply build.gradle]: Could be a separate pipeline maybe

**Comment by jaredsburrows** [comment]
@shashachu I actually removed that plugin here: https://github.com/pinterest/ktlint/pull/465/files#diff-6bb5647686cc692dfa23d5d48d46e32aL1.

If you run `gradlew build`, then look in `./ktlint/ktlint/build/distributions`, you will see `ktlint-0.3.4-SNAPSHOT.zip`. If you unzip it, `unzip ktlint-0.3.4-SNAPSHOT.zip`, you will see the `bin` and `lib` folder. The `bin` folder has the `ktlint` bash script. Then you run `./ktlint --help`

**Comment by jaredsburrows** [comment]
@shashachu If you want this functionality back, I think you should follow the setup here: https://github.com/pinterest/ktlint/pull/465/files#diff-6bb5647686cc692dfa23d5d48d46e32aL1.

**Comment by shashachu** [comment]
We don't want to use the bash distribution, as it requires the `lib` folder and all the separate jar dependencies. I think the executable jar is the right approach. Sorry, I didn't notice the plugin had been removed in that diff.

**Comment by shashachu** [comment]
> @shashachu If you want this functionality back, I think you should follow the setup here: https://github.com/pinterest/ktlint/pull/465/files#diff-6bb5647686cc692dfa23d5d48d46e32aL1.

I had originally tried the same setup and it didn't actually include any of the jars built from the sub-modules. I can give it another stab this weekend.

**Comment by jaredsburrows** [comment]
I have added the plugin to my PR here: https://github.com/pinterest/ktlint/pull/471.

**Comment by Tapchicoma** [comment]
My 5 cents here: @shashachu if you will use Gradle for next release publishing, could you ensure that `ktlint` executable behaves the same as if it was built with maven?

My point that it should work without any additional jars in the class path and run by simply calling `./ktlint`

**Comment by shashachu** [comment]
> My 5 cents here: @shashachu if you will use Gradle for next release publishing, could you ensure that `ktlint` executable behaves the same as if it was built with maven?
> 
> My point that it should work without any additional jars in the class path and run by simply calling `./ktlint`

Totally agree. And there's no reason why the Gradle build needs to be tied to any particular release since it's used internally. Whenever we do make the switch, we will make sure the behavior is identical to the existing maven build.

File: build.gradle [hunk 2]
```diff
@@ -1,5 +1,7 @@
 plugins {
   id 'org.jetbrains.kotlin.jvm' version '1.3.31' apply false
+  id 'com.github.johnrengelman.shadow' version '5.0.0'
+  id 'java'
```
Comment (JLLeitschuh) [inline build.gradle]: Consider using the `java-library` plugin instead. The `java` plugin is legacy.

**Pull request closed without merging**